fineract-investor/src/main/java/org/apache/fineract/investor/service/LoanAccountOwnerTransferServiceImpl.java

```diff
@@ -0,0 +1,210 @@
+
+    private void declinePendingSaleTransfer(Loan loan, ExternalAssetOwnerTransfer pendingSaleTransfer, LocalDate settlementDate) {
+        ExternalAssetOwnerTransfer declinedSaleTransfer = createDeclinedTransfer(pendingSaleTransfer, settlementDate, loan);
+        updatePendingTransfer(pendingSaleTransfer, settlementDate);
+
+        businessEventNotifierService.notifyPostBusinessEvent(new LoanOwnershipTransferBusinessEvent(declinedSaleTransfer, loan));
+    }
+
+    private void executePendingBuybackTransfer(final Loan loan, ExternalAssetOwnerTransfer buybackTransfer,
+            final LocalDate settlementDate) {
+        ExternalAssetOwnerTransfer activeTransfer = findActiveTransfer(loan, buybackTransfer);
+        updateActiveTransfer(activeTransfer, settlementDate);
+        buybackTransfer = updatePendingBuybackTransfer(loan, buybackTransfer, settlementDate);
+
+        externalAssetOwnerTransferLoanMappingRepository.deleteByLoanIdAndOwnerTransfer(loan.getId(), activeTransfer);
+        accountingService.createJournalEntriesForBuybackAssetTransfer(loan, buybackTransfer);
+
+        businessEventNotifierService.notifyPostBusinessEvent(new LoanOwnershipTransferBusinessEvent(buybackTransfer, loan));
+        businessEventNotifierService.notifyPostBusinessEvent(new LoanAccountSnapshotBusinessEvent(loan));
+    }
+
+    private ExternalAssetOwnerTransfer createCancelledTransfer(ExternalAssetOwnerTransfer pendingTransfer, LocalDate settlementDate,
+            ExternalTransferSubStatus subStatus) {
+        ExternalAssetOwnerTransfer cancelledTransfer = new ExternalAssetOwnerTransfer();
+        cancelledTransfer.setOwner(pendingTransfer.getOwner());
+        cancelledTransfer.setExternalId(pendingTransfer.getExternalId());
+        cancelledTransfer.setStatus(CANCELLED);
+        cancelledTransfer.setSubStatus(subStatus);
+        cancelledTransfer.setSettlementDate(settlementDate);
+        cancelledTransfer.setLoanId(pendingTransfer.getLoanId());
+        cancelledTransfer.setExternalLoanId(pendingTransfer.getExternalLoanId());
+        cancelledTransfer.setPurchasePriceRatio(pendingTransfer.getPurchasePriceRatio());
+        cancelledTransfer.setEffectiveDateFrom(settlementDate);
+        cancelledTransfer.setEffectiveDateTo(settlementDate);
+        return externalAssetOwnerTransferRepository.save(cancelledTransfer);
+    }
+
+    private ExternalAssetOwnerTransfer createDeclinedTransfer(ExternalAssetOwnerTransfer pendingSaleTransfer, LocalDate settlementDate,
+            Loan loan) {
+        ExternalAssetOwnerTransfer declinedTransfer = new ExternalAssetOwnerTransfer();
+        declinedTransfer.setOwner(pendingSaleTransfer.getOwner());
+        declinedTransfer.setExternalId(pendingSaleTransfer.getExternalId());
+        declinedTransfer.setStatus(DECLINED);
+        declinedTransfer.setSubStatus(isBiggerThanZero(loan.getTotalOverpaid()) ? BALANCE_NEGATIVE : BALANCE_ZERO);
+        declinedTransfer.setSettlementDate(settlementDate);
```

Comment:
please use the original settlement date

Reply:
Okay.
Question: is this the case for the new `CANCELLED` transfers as well? Should I use the original settlement date for them, or the current businessDate?